### PR TITLE
Mutation Observer performance

### PR DESCRIPTION
**Pull request by alexanderby**
- Scan the whole DOM tree for changes when there's too large count of mutations.
- Don't handle unnecessary mutations.
- Another attempt for #535.

### CODE REVIEW

**Comment by Gusted** [comment]
Poor CPU & Dark Reader(mutation-observer-performance branch) Vs big DOM website https://www.w3.org/TR/2018/WD-html53-20181018/

 Profiler link: https://perfht.ml/3dYYOat

**Comment by Gusted** [comment]
Poor CPU & Dark Reader(mutation-observer-performance branch) Vs big DOM website https://www.w3.org/TR/2018/WD-html53-20181018/

https://perfht.ml/36c7urv

File: src/inject/utils/dom.ts [hunk 1]
```diff
@@ -145,3 +145,153 @@ export function iterateShadowNodes(root: Node, iterator: (node: Element) => void
...
+
+const HUGE_MUTATIONS_COUNT = 1000;
+
+function isHugeMutation(mutations: MutationRecord[]) {
+    if (mutations.length > HUGE_MUTATIONS_COUNT) {
+        return true;
+    }
+
+    let addedNodesCount = 0;
+    for (let i = 0; i < mutations.length; i++) {
```
Comment (Gusted) [inline src/inject/utils/dom.ts:182]: Length caching

**Comment by Gusted** [comment]
Poor CPU & Dark Reader(36a43b5) Vs big DOM website 
https://www.w3.org/TR/2018/WD-html53-20181018/

And this website is resolved 3 seconds until Interactive and some seconds later not more loading icon :D


https://perfht.ml/2WKnqy5

File: src/inject/utils/dom.ts (hunk 1 above)
Comment (alexanderby) [reply src/inject/utils/dom.ts:182]: Let's trash what? What's the point of length caching?

Comment (Gusted) [reply src/inject/utils/dom.ts:182]: It was a joke :) 
But 
https://jsben.ch/kidOp 
Just that little performance boost

Comment (alexanderby) [reply src/inject/utils/dom.ts:182]: The performance boost can't be little :)

Here's what I see in Firefox
Screen Shot 2020-05-19 at 19 38 03
and after a 3rd relaunch in Chrome
Screen Shot 2020-05-19 at 19 37 11
So, it's obvious that `for..of` is 4 times slower in Firefox and 25% slower in Chrome. But all other functions are equal, and I choose what looks clearer (though some people prefer to use `Array.some` or `Array.every` to exit loop earlier, I think `for` loop is more clear).

Comment (Gusted) [reply src/inject/utils/dom.ts:182]: I don't get at all the same result........
Poor laptop

Comment (alexanderby) [reply src/inject/utils/dom.ts:182]: ... And maybe you've seen this saying of Donald Knuth
> Programmers waste enormous amounts of time thinking about, or worrying about, the speed of noncritical parts of their programs, and these attempts at efficiency actually have a strong negative impact when debugging and maintenance are considered. We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%

Comment (Gusted) [reply src/inject/utils/dom.ts:182]: If every time someone says `BUT THIS DROP PERFORMANCE` etc it start to get worried and I didn't though about performance at all if you look at my earliest PR's for darkreader with changes of codebase I used var =. I used a lot of functions what easy could be reduced to 1 etc.

**Pull request merged**